[MLIR][OpenMP] Add a new ComposableOpInterface to check/set a composite unitAttr.

TIFitis opened the pull request:
Adds a new ComposableOpInterface for OpenMP operations that can represent a single leaf of a composite OpenMP construct.

TIFitis committed:
[MLIR][OpenMP] Add a new ComposableLoopWrapperInterface to check and set a composite unitAttr.

Meinersbur commented:
Can you elaborate how it it useful? E.g. using a test case.

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:
```diff
@@ -129,6 +129,7 @@ def PrivateClauseOp : OpenMP_Op<"private", [IsolatedFromAbove, RecipeInterface]>
 def ParallelOp : OpenMP_Op<"parallel", traits = [
     AttrSizedOperandSegments, AutomaticAllocationScope,
     DeclareOpInterfaceMethods<LoopWrapperInterface>,
+    DeclareOpInterfaceMethods<ComposableLoopWrapperInterface>,
```
Nit: This should go above the previous trait, since they are alphabetically sorted. Same comment for the other additions to this file.

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:
```diff
@@ -142,6 +142,40 @@ def LoopWrapperInterface : OpInterface<"LoopWrapperInterface"> {
   ];
 }
 
+def ComposableLoopWrapperInterface : OpInterface<"ComposableLoopWrapperInterface"> {
```
I think `LoopWrapper` shouldn't be part of this interface's name, because it would still be applied to `omp.parallel` after we remove it from the list of loop wrapper operations. I can suggest `ComposableOpInterface` or something similar.

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:
```diff
@@ -142,6 +142,40 @@ def LoopWrapperInterface : OpInterface<"LoopWrapperInterface"> {
 
+def ComposableLoopWrapperInterface : OpInterface<"ComposableLoopWrapperInterface"> {
+  let description = [{
+    OpenMP operations that can wrap a single loop nest. When taking a wrapper
+    role, these operations must only contain a single region with a single block
+    in which there's a single operation and a terminator. That nested operation
+    must be another loop wrapper or an `omp.loop_nest`.
```
This description needs updating. It should say that this is for OpenMP operations that can represent a single leaf of a composite OpenMP construct. It would also be good to use that description to update this PR's description, so it's easier for reviewers to understand what this change is for.

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:
```diff
@@ -142,6 +142,40 @@ def LoopWrapperInterface : OpInterface<"LoopWrapperInterface"> {
+  let methods = [
+    InterfaceMethod<
+      /*description=*/[{
+        Check whether the operation is composable.
```
Feel free to reword it as you see fit, the important thing is that the presence of the attribute represents that the operation is representing one leaf of a composite construct.
```suggestion
        Check whether the operation is representing a leaf of a composite OpenMP construct.
```

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:
```diff
@@ -142,6 +142,40 @@ def LoopWrapperInterface : OpInterface<"LoopWrapperInterface"> {
+      /*retTy=*/"void",
+      /*methodName=*/"setComposite",
+      (ins), [{}], [{
+        $_op->setDiscardableAttr("omp.composite", mlir::UnitAttr::get($_op->getContext()));
```
I think this method should take a boolean argument. If it's `false`, it should remove the attribute if present.

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:
```diff
@@ -142,6 +142,40 @@ def LoopWrapperInterface : OpInterface<"LoopWrapperInterface"> {
+    >,
+    InterfaceMethod<
+      /*description=*/[{
+        Set the CompositeAttr for the op.
```
```suggestion
        Mark the operation as part of an OpenMP composite construct.
```

skatrak reviewed (commented):
Thank you Akash for working on this, I've got a few comments that should be straightforward to address.

skatrak commented:
> Can you elaborate how it it useful? E.g. using a test case.

I agree that this PR needs a description. It's probably not possible to add a test case just for this change (this would be easier to test when uses for it are added to e.g. Flang lowering).

TIFitis committed:
Address reviewer comments.

TIFitis commented:
Thanks for the review @skatrak I've addressed the comments in the latest revision.

@Meinersbur I've updated the description both in code comment and PR to give a better idea. There's no way to test the OpInterfaces AFAIK, or at least the other OpInterfaces in the same file don't have any exclusive tests for them.
I will however be adding a couple of PRs shortly which makes use of this new OpInterface, through that we can have some testing for this as well.

Thanks,
Akash

TIFitis commented:
I've moved this PR to #102340 to have a series of patches that make use of this work.
